### PR TITLE
[HWASan] [compiler-rt] support non-4k pages on Android

### PR DESCRIPTION
**Pull request by fmayer**

### CODE REVIEW

File: compiler-rt/lib/hwasan/hwasan_dynamic_shadow.cpp [hunk 1]
```diff
@@ -36,15 +36,18 @@ decltype(__hwasan_shadow)* __hwasan_premap_shadow();
 
 namespace __hwasan {
 
+// We cannot call anything in libc here (see comment above), so we need to
+// assume the biggest allowed page size.
+static constexpr uptr kMaxGranularity = 16384;
```
Comment (pinskia) [inline compiler-rt/lib/hwasan/hwasan_dynamic_shadow.cpp:41]: Won't it be 64k rather than 16k? Since that is the max page size for aarch64 linux?

Comment (fmayer) [reply compiler-rt/lib/hwasan/hwasan_dynamic_shadow.cpp:41]: android max page size is 16k, defined here: https://android.googlesource.com/platform/bionic/+/main/libc/platform/bionic/page.h#41

File: compiler-rt/lib/hwasan/hwasan_dynamic_shadow.cpp [hunk 2]
```diff
@@ -109,7 +114,7 @@ uptr FindDynamicShadowStart(uptr shadow_size_bytes) {
   if (IsPremapShadowAvailable())
     return FindPremappedShadowStart(shadow_size_bytes);
   return MapDynamicShadow(shadow_size_bytes, kShadowScale, kShadowBaseAlignment,
-                          kHighMemEnd);
+                          kHighMemEnd, kMaxGranularity);
```
Comment (eugenis) [inline compiler-rt/lib/hwasan/hwasan_dynamic_shadow.cpp:117]: Why can't this be GetMmapGranularity?

File: compiler-rt/lib/sanitizer_common/sanitizer_common.h [hunk 3]
```diff
@@ -139,7 +137,8 @@ void UnmapFromTo(uptr from, uptr to);
 // shadow_size_bytes bytes on the right, which on linux is mapped no access.
 // The high_mem_end may be updated if the original shadow size doesn't fit.
 uptr MapDynamicShadow(uptr shadow_size_bytes, uptr shadow_scale,
-                      uptr min_shadow_base_alignment, uptr &high_mem_end);
+                      uptr min_shadow_base_alignment, uptr &high_mem_end,
+                      const uptr granularity);
```
Comment (eugenis) [inline compiler-rt/lib/sanitizer_common/sanitizer_common.h:141]: We never put const on non-pointer non-reference function arguments.

**Comment by fmayer** [comment]
Ping.

**Review by eugenis — APPROVED** [review]

**Comment by nico** [comment]
This breaks building on windows: http://45.33.8.238/win/89650/step_3.txt

Please take a look and revert for now if it takes a while to fix.

**Comment by fmayer** [comment]
Just reverted, sorry I didn't revert earlier because lab.llvm.org was down,
so I couldn't confirm what happened.

**Pull request merged**